---
title: "Insectary_comp_1-decontam"
author: "laura_brettell"
date: "16th April 2024"
output: html_document
---

# The project

Does rearing in a different insectary affect the mosquito microbiome?

# The setup

one cohort of Ae aegypti eggs were split into 3 and reared in standard conditions to see how their microbiomes varied between insectaries at difff life stages.

# install packages

```{r}
# phyloseq itself
#if(!requireNamespace("BiocManager")){
#  install.packages("BiocManager")
#   }
#BiocManager::install("phyloseq")

# qiime2r
# to be able to convert my qiime2 files
#if (!requireNamespace("devtools", quietly = TRUE)){install.packages("devtools")}
# devtools::install_github("jbisanz/qiime2R") # current version is 0.99.20
 
# decontam for filtering contaminant reads
# if (!require("BiocManager", quietly = TRUE))
#       install.packages("BiocManager")
# BiocManager::install("decontam")


library(phyloseq)
library(qiime2R)
library(ggplot2)
#library(tidyverse)
library(dplyr) # for select command
library(decontam)

```
# import data and create phyloseq object

Take data from qiime2 and put it into a phyloseq object that decontam can work with

https://vaulot.github.io/tutorials/Phyloseq_tutorial.html
https://bioconductor.org/packages/release/bioc/vignettes/phyloseq/inst/doc/phyloseq-basics.html


The qiime2 part was done by novogene:

"Raw data were subjected to denoising using DADA2 (no additional trimming) and taxonomy was assigned to ASVs by applying the classify-sklearn algorithm in Qiime2 (202202) using a Naïve Bayes classifier pre-trained in the SILVA 138.1 database. The phylogenetic relationships between ASVs were determined in qiime2 through a multiple sequence alignment using MAFFT and phylogenetic reconstruction using fasttree.

```{r}
# setwd("C:/Users/SES499/OneDrive - University of Salford/Desktop/my_files/PROJECTS/Insectary_Comp/data/Analysis_Dec2023/")


# import taxonomy 
tax <- read_qza("seq_taxonomy.qza")$data # read in data 'Taxonomy annotation results (qza format)'
tax <- parse_taxonomy(tax) # make in more useful format
tax <- as.matrix(tax) # must be matrix for later steps
#rownames(tax) # checking
#colnames(tax) # more checking
#class(tax) # rechecking is matrix
TAX = phyloseq::tax_table(tax) # convert to necessary format (tax_table) to make phyloseq object
# I had to specify it was the phyloseq command I was after. The package MicrobiotaProcess has the command but with a differetn function
#TAX # checking

# import phylogeny
tre1 <- read_tree("tree.nwk")


# import data
total_table <- read.delim("featureTable.sample.total.absolute.txt") #'feature table of absolute abundances'
rownames(total_table) <- total_table$X.OTU_num # set rownames
total_table <- subset(total_table, select = -X.OTU_num) # remove extra col

# reduce to just ASVs
ASVs <- subset(total_table, select = -Taxonomy)
ASVs <- as.matrix(ASVs) # convert to matrix
#rownames(ASVmat) # chekcing the same as tax
#colnames(ASVmat) # checking
#class(ASVmat) # again must be matrix
OTU = otu_table(ASVs, taxa_are_rows = TRUE) # convert to necessary format (matrix) to make phyloseq object
#OTU # checking


# create phyloseq object
physeq = phyloseq(OTU, TAX) #create phyloseq object
#physeq # look at it
#sample_names(physeq) # check
#temp <- as.data.frame(tax_table(physeq)) # checking taxa columns are right



# metadata
metadata <- read_q2metadata("insectary_comp_full_samplelist3.txt") # metadata
metadata <- subset(metadata, select = c(SampleID, Sample_Type2, for_ordering, Building, Replicate, group_for_novogene, sample_or_neg)) # reduce to useful bits
rownames(metadata) <- metadata$SampleID

samples = sample_data(metadata)

physeq <- phyloseq(OTU, samples, TAX)
# add phylogeny info
physeq <- merge_phyloseq(physeq, tre1)

```

# Decontam

## first I'm going to remove NA's (the arabiensis samples that aren't part of this study)

```{r}
physeq <- prune_samples(sample_data(physeq)$sample_or_neg != "negative", physeq)
       
```

```{r}
df <- as.data.frame(sample_data(physeq)) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(physeq)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
ggplot(data=df, aes(x=Index, y=LibrarySize, color=sample_or_neg))  + geom_point(alpha=0.7, size=1)
       
```


# identify contaminants using Decontam

Using the “prevalence” method. In this method, the prevalence (presence/absence across samples) of each sequence feature in true positive samples is compared to the prevalence in negative controls to identify contaminants.

The default for determining a contaminant is that it reaches a probability of 0.1 in the statistical test being performed. Here I am using a threshold=0.5 (as suggested by Diaz et al 2021), that will identify as contaminants all sequences that are more prevalent in negative controls than in positive samples. 


```{r}
sample_data(physeq)$is.neg <- sample_data(physeq)$sample_or_neg == "neg"
# summarizing the data on whether each sample is a true sample or control as a logical variable, with TRUE for control samples, as that is the form required by isContaminant.
contamdf.prev05 <- isContaminant(physeq, method="prevalence", neg="is.neg", threshold = 0.5)
table(contamdf.prev05$contaminant)
```

This has assigned 706 ASVs as contaminants and 22606 ASVs as 'real'.

Now inspect the output by plotting all ASVs, whether decontam has called them contaminants or not and their prevalence in true vs control samples (and check it looks appropriate)

```{r}
# Make phyloseq object of presence-absence in negative controls and true samples
physeq.pa <- transform_sample_counts(physeq, function(abund) 1*(abund>0))
physeq.pa.neg <- prune_samples(sample_data(physeq.pa)$sample_or_neg == "neg", physeq.pa)
physeq.pa.pos <- prune_samples(sample_data(physeq.pa)$sample_or_neg == "sample", physeq.pa)
# Make data.frame of prevalence in positive and negative samples
df.pa <- data.frame(pa.pos=taxa_sums(physeq.pa.pos), pa.neg=taxa_sums(physeq.pa.neg),
                      contaminant=contamdf.prev05$contaminant)
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (True Samples)")

```

yep, looks appropriate

```{r}

#now to remove the ASVs determined to be contaminants
physeq_cont_removed <- prune_taxa(!contamdf.prev05$contaminant, physeq)
# now remove negative controls and keep only the samples
physeq_cont_removed <- prune_samples(sample_data(physeq_cont_removed)$sample_or_neg == "sample", physeq_cont_removed)

temp <- as.data.frame(phyloseq::tax_table(physeq_cont_removed))

# remove chloroplast seqs
physeq_cont_removed <- subset_taxa(physeq_cont_removed, (Order != "Chloroplast") | is.na(Class))

# check they've gone - yes
#temp <- as.data.frame(phyloseq::tax_table(physeq_cont_removed))
# reduced to 20903 ASVs

# remove mitochondrial seqs
physeq_cont_removed <- subset_taxa(physeq_cont_removed, (Family != "Mitochondria") | is.na(Family))
#temp <- as.data.frame(phyloseq::tax_table(physeq_cont_removed))
# reduced to 20690 ASVs

# keep only bacteria
physeq_cont_removed <- subset_taxa(physeq_cont_removed, (Kingdom == "Bacteria"))

# check
#temp3 <- as.data.frame(phyloseq::tax_table(physeq_cont_removed))
# reduced to 20065 ASVs
# extra manual check  - looks good
#write.csv(temp, "ASVs_after_filtering.csv")

```


# Create an object containing only contaminants

I want a record of which ASVs were assigned as contaminants and removed, for the supp data.

```{r}
physeq_contaminants <- prune_taxa(contamdf.prev05$contaminant, physeq)

contaminant_list <-phyloseq::tax_table(physeq_contaminants)

#write.csv(contaminant_list, "contaminant_ASVs_by_decontam.csv")
#these dont contain mito, chloro, etc, (well they might also have some of these) just those identified by decontam.

```

# save phyloseq object

```{r}
saveRDS(physeq_cont_removed, "physeq_cont_removed")

```


## get a file of read lengths of samples passed QC



```{r}
df2 <- as.data.frame(sample_data(physeq_cont_removed)) # Put sample_data into a ggplot-friendly data.frame
df2$LibrarySize <- sample_sums(physeq_cont_removed)

# save for paper supp
df3 <- as.matrix(df2) # for some reason I need to do this
#write.csv(df3, "metadata_with_read_depths.csv")


    
```



