---
title: "Ins_comp_5-ASVs"
author: "laura"
date: "2024-01-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# The project

Does rearing in a different insectary affect the mosquito microbiome?

# The setup

one cohort of Ae aegypti eggs were split into 3 and reared in standard conditions to see how their microbiomes varied between insectaries at diff life stages.


Now I will try a couple of ways to investigate whether there are signatures or particular ASVs associated with different insectaries and in which sample types these appear - is it larvae and larval water as suggested from diversity plot groupings or is it others, or do signature ASVs persist?

# Install packages

```{r}
# phyloseq itself
#if(!requireNamespace("BiocManager")){
#  install.packages("BiocManager")
#   }
#BiocManager::install("phyloseq")

# if (!requireNamespace("BiocManager", quietly=TRUE))
 #   install.packages("BiocManager")
## BiocManager::install("BiocUpgrade") ## you may need this
#BiocManager::install("MicrobiotaProcess")
 

#install.packages("gghalves") # required for `mp_plot_alpha()`
#install.packages("ggh4x") # required for `mp_plot_alpha()`


library(phyloseq)
library(ggplot2)
#library(tidyverse)
library(dplyr) # for select command

```

# import data

```{r}

rm(list = ls()) # clear env
ps = readRDS("physeq_obj_3974") # currently phyloseq obj
head(sample_data(ps))

```

# first - heatmap using same method as genus box plot

```{r}
asv_abundance <- ps %>%
  #tax_glom(taxrank = "OTU") %>%                        # Set to smallest taxonomic level you are interested in
  transform_sample_counts(function(x) {x/sum(x)} ) %>%   # Transform to rel. abundance
  psmelt()                                               # Melt to long format
head(asv_abundance)

```

```{r}
all <- asv_abundance %>%
  select(Phylum, Class, Family, Genus, OTU, SampleID, Abundance, Sample_Type2, Building) %>%
  filter(Abundance != 0) %>%
  mutate(
    Phylum = as.character(Phylum),
    Class = as.character(Class),
    Family = as.character(Family),
    Genus = as.character(Genus),
    SampleID = as.character(SampleID))
head(all)

```


```{r}

#ASV_subset <- as.matrix(subset(all, all$Abundance >= 0.05))
#ASV_subset %>%
#  as.numeric(as.double(ASV_subset$Abundance)) 
asv_subset <- subset(all, all$Abundance >= 0.05)
head(asv_subset)

is.na(asv_subset) %>% table()


```
There are NAs so won't work, event hough these are in genus etc I think - so make all NAs into 'unknown'
?

```{r}
#library(tidyverse)
#df <- asv_subset %>%
#    replace_na(list(Genus = 'unknown'))
#is.na(df) %>% table()

#df <- df %>%
#    replace_na(list(Family = 'unknown'))
#is.na(df) %>% table()

#df <- df %>%
#    replace_na(list(Class = 'unknown'))
#is.na(df) %>% table()


#df <- df %>%
#    replace_na(list(Phylum = 'unknown'))
#is.na(df) %>% table()

# now no NAs remain

#df <- as.matrix(df)

```


```{r}
library(pheatmap)
#pheatmap(df)


```
still won't work for some reason, saying there's NAs when it desn't look there are = who knows.    


# method 2 - following phyloseq demo

of this fails i will deconstruct my phyloseq obj an dfilter then recreate folloing the isofemale method


https://joey711.github.io/phyloseq-demo/phyloseq-demo-slides.html
also useful : https://microucph.github.io/amplicon_data_analysis/html/phyloseq_operations.html
https://github.com/joey711/phyloseq/issues/1089



```{r}

# transform counts to rel abund
ps2 <- transform_sample_counts(ps, function(x) x/sum(x))

#Remove taxa not seen more than 0.05 rel abund in at least 0.004% of the samples (this equals 1 sample as there are 244 in total). This protects against an OTU with small mean & trivially large C.V. (I can't work out how to say - in any one sample)

ps3 = filter_taxa(ps2, function(x) sum(x >= 0.05) >= (0.004*length(x)), TRUE) # this has 34 ASVs



```



```{r}

heatmap_test <- plot_heatmap(
              ps3, 
              method = NULL,
              taxa.label = "Family",
             taxa.order = "Family",
             sample.order = "Sample_Type2",
             sample.label = "Sample_Type2",
             trans = NULL,
             low = "white",
             high = "darkblue",
             na.value = "white",
             title = "heatmap")

heatmap_test


```

good stuff, now to improve visualisation



```{r}
heatmap_test2 <- plot_heatmap(
              ps3, 
              method = NULL,
              taxa.label = "Genus",
             taxa.order = "Family",
             sample.order = "for_ordering",
             sample.label = "Sample_Type2",
             trans = NULL,
            low = "lightblue",
            high = "darkblue",
        #     na.value = "white",
             title = "relative abundance of all ASVs seen at >= 5% reads in >= 1 sample")

heatmap_test2


```




```{r}
heatmap_test3 <- heatmap_test2 + facet_wrap(~Building, scales = "free_x", nrow = 1)

heatmap_test3

```

### trying again to sort out the zeros/NAs


```{r}
# take components out of phloseq obj

tax <- tax_table(ps3)
meta <- sample_data(ps3)
asv <- otu_table(ps3)

asvdf <- as.data.frame(asv)
asvdf[asvdf == 0] <- NA 
  

asv2 <- as.matrix(asvdf) # convert to matrix
asv2 = otu_table(asv2, taxa_are_rows = TRUE) # convert to necessary format  to make phyloseq object


psNAs <- phyloseq(asv2, tax, meta)  

```

```{r}
heatmap_test4 <- plot_heatmap(
              psNAs, 
              method = NULL,
              taxa.label = "Genus",
             taxa.order = "Family",
             sample.order = "for_ordering",
             sample.label = "Sample_Type2",
             trans = NULL,
            low = "#B1C9E8",
            high = "#002F6C",
             na.value = "lightgrey",
             title = "relative abundance of all ASVs seen at >= 5% reads in >= 1 sample")

heatmap_test4


```
```{r}
heatmap_test5 <- heatmap_test4 + facet_wrap(~Building, scales = "free_x", nrow = 1)

heatmap_test5

```
now facet by sample type which is probs better
```{r}
heatmap_test6 <- plot_heatmap(
              psNAs, 
              method = NULL,
              taxa.label = "Genus",
             taxa.order = "Family",
             sample.order = "for_ordering",
             sample.label = "Building",
             trans = NULL,
            low = "#B1C9E8",
            high = "#002F6C",
             na.value = "lightgrey",
             title = "relative abundance of all ASVs seen at >= 5% reads in >= 1 sample")

heatmap_test6


```

```{r}
heatmap_test7 <- heatmap_test6 + facet_grid(~Sample_Type2, scales = "free_x", space = "free_x")

heatmap_test7 

#heatmap_test7 <- heatmap_test6 +  ggforce::facet_col(vars(),
#             space="free",
#             scale="free_y")

```

```{r}


```